src/main/java/socket/BidWS.java

```diff
@@ -353,6 +379,155 @@ protected void onAuctionBid(BodyWS body) {
+        }
+
+        if (!dbAuction.status.equals(Auction.IN_PROGRESS)) {
+            String json = JsonCommon.error(AUCTION_NOT_IN_PROGRESS);
+            sender.reply(session, body, BodyWSCommon.error(json));
+            return;
+        }
+
+        // 3. User checks
+        User dbUser;
+        try {
+            dbUser = userDAO.getById(userId);
+        } catch (DAOException e) {
+            Logger.error("Bid get user by ID at combinatorial auction", String.valueOf(userId), e.toString());
+            sender.reply(session, body, BodyWSCommon.internalServerError());
+            return;
+        }
+
+        // 3.1. User hasn't any bid on this auction
+        Set<Bid> userBids;
+        try {
+            userBids = new HashSet<>(bidDAO.getListByOwnerId(userId));
+        } catch (DAOException e) {
+            Logger.error("Bid get user bids at combinatorial auction", String.valueOf(userId), e.toString());
+            sender.reply(session, body, BodyWSCommon.internalServerError());
+            return;
+        }
+        Set<Bid> auctionBids;
+        try {
+            auctionBids = new HashSet<>(bidDAO.getListByAuctionId(commonAuctionId));
+        } catch (DAOException e) {
+            Logger.error("Bid get auction bids at combinatorial auction", String.valueOf(commonAuctionId), e.toString());
+            sender.reply(session, body, BodyWSCommon.internalServerError());
+            return;
+        }
+        userBids.retainAll(auctionBids);
+        if (userBids.size() != 0) {
+            String json = JsonCommon.error(USER_ALREADY_BIDDED);
+            sender.reply(session, body, BodyWSCommon.error(json));
+            return;
+        }
+
+        // 3.2. User has enough credit
+        if (dbUser.credit < commonAmount) {
+            String json = JsonCommon.error(NO_CREDIT);
+            sender.reply(session, body, BodyWSCommon.error(json));
+            return;
+        }
+
+        // 3.3. Bid amount is enough to bid for the auction
+        if (commonAmount < dbAuction.startingPrice) {
+            String json = JsonCommon.error(LOW_BID_STARTING_PRICE);
+            sender.reply(session, body, BodyWSCommon.error(json));
+            return;
+        }
+
+        // Al checks passed ( @zurfyx likes this part ;) )
+        Bid dbBid = null;
+        for (Bid unsafeBid: unsafeBidsList) {
+            Bid newBid = new Bid();
+            newBid.amount = unsafeBid.amount;
+            newBid.auctionId = unsafeBid.auctionId;
+            newBid.goodId = unsafeBid.goodId;
+            newBid.ownerId = this.httpSession.userId();
+
+            try {
+                dbBid = bidDAO.create(newBid);
+            } catch (DAOException e) {
+                Logger.error("Create bid", newBid.toString(), e.toString());
+                sender.reply(session, body, BodyWSCommon.internalServerError());
+                return;
+            }
+        }
+
+        // In this case we protect the bid amount
+        dbBid.amount = 0.0;
```

Comment:
If you're doing this here  you also need to restrict all servlets transactions listings.

Reply:
True. I will modify this later.